pkg/clusteragent/admission/validate/kubernetesadmissionevents/kubernetesadmissionevents.go

```diff
@@ -0,0 +1,199 @@
+		matchConditions: []admissionregistrationv1.MatchCondition{
+			{
+				Name:       "exclude-system-users",
+				Expression: "!(request.userInfo.username.startsWith('system:'))",
+			},
+		},
+		demultiplexer:           demultiplexer,
+		supportsMatchConditions: supportsMatchConditions,
+		checkid:                 "kubernetes_admission_events",
+	}
+}
+
+// Name returns the name of the webhook
+func (w *Webhook) Name() string {
+	return w.name
+}
+
+// WebhookType returns the type of the webhook
+func (w *Webhook) WebhookType() common.WebhookType {
+	return common.ValidatingWebhook
+}
+
+// IsEnabled returns whether the webhook is enabled
+func (w *Webhook) IsEnabled() bool {
+	return w.isEnabled
+}
+
+// Endpoint returns the endpoint of the webhook
+func (w *Webhook) Endpoint() string {
+	return w.endpoint
+}
+
+// Resources returns the kubernetes resources for which the webhook should
+// be invoked
+func (w *Webhook) Resources() map[string][]string {
+	return w.resources
+}
+
+// Operations returns the operations on the resources specified for which
+// the webhook should be invoked
+func (w *Webhook) Operations() []admissionregistrationv1.OperationType {
+	return w.operations
+}
+
+// LabelSelectors returns the label selectors that specify when the webhook
+// should be invoked
+func (w *Webhook) LabelSelectors(useNamespaceSelector bool) (namespaceSelector *metav1.LabelSelector, objectSelector *metav1.LabelSelector) {
+	return common.DefaultLabelSelectors(useNamespaceSelector)
+}
+
+// MatchConditions returns the Match Conditions used for fine-grained
+// request filtering
+func (w *Webhook) MatchConditions() []admissionregistrationv1.MatchCondition {
+	return w.matchConditions
+}
+
+// WebhookFunc returns the function that generates a Datadog Event and admits the request.
+func (w *Webhook) WebhookFunc() admission.WebhookFunc {
+	return func(request *admission.Request) *admiv1.AdmissionResponse {
+		return common.ValidationResponse(validatecommon.Validate(request, w.Name(), w.emitEvent, request.DynamicClient))
+	}
+}
+
+// emitEvent generates a Datadog Event, sends it to the kubernetes_admission_events sender and admits the request.
+func (w *Webhook) emitEvent(request *admission.Request, _ string, _ dynamic.Interface) (bool, error) {
+	if !w.supportsMatchConditions {
+		// Manually filter out system users if match conditions are not supported.
+		if strings.HasPrefix(request.UserInfo.Username, "system:") {
+			log.Debugf("Skipping system user: %s", request.UserInfo.Username)
+			return true, nil
+		}
+	}
+
+	// Decode object and oldObject.
+	var newResource unstructured.Unstructured
+	if request.Operation != admissionregistrationv1.Delete {
+		if err := json.Unmarshal(request.Object, &newResource); err != nil {
+			return true, fmt.Errorf("failed to unmarshal object: %w", err)
+		}
+	}
+	var oldResource unstructured.Unstructured
+	if request.Operation != "CREATE" && request.Operation != "CONNECT" {
```

Comment:
I was a bit worried as some point about using `admissionregistrationv1` directly in the webhook as the code is the same between `v1` and `v1beta1`. But since behind it's just a string there should be no issues.
Done: https://github.com/DataDog/datadog-agent/pull/30715/commits/68d181f4f5833289b3494842476ff2f698d23651